File: ExampleApp/src/MapGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Prism;
namespace Example
{
    public class MapGenerator : Behaviour
    {
        // [EditorSlider("MapWidth Custom Name", 2, 0, 1024)]
        public int MapWidth = 128;
        public int MapHeight = 128;
        public int Octaves = 4;
        public float Persistance = 0.74f;
        public int Seed = 21;
        public float Lacunarity = 3.0f;
        public Vector2 Offset = new Vector2(13.4f, 6.26f);
        public float NoiseScale = 0.5f;

        public float Speed = 0.0f;
        public void GenerateMap()
        {
            //float[,] noiseMap = Noise.GenerateNoiseMap(mapWidth, mapHeight, noiseScale);
            float[,] noiseMap = Noise.GenerateNoiseMap(MapWidth, MapHeight, Seed, NoiseScale, Octaves, Persistance, Lacunarity, Offset);

            uint width = (uint)noiseMap.GetLength(0);
            uint height = (uint)noiseMap.GetLength(1);

            Texture2D texture = new Texture2D(width, height);
            Vector4[] colorMap = new Vector4[width * height];
            for (int y = 0; y < height; y++)
            {
                for (int x = 0; x < width; x++)
                {
                    colorMap[x + y * width] = Vector4.Lerp(Color.Black, Color.White, noiseMap[x, y]);
                }
            }

            texture.SetData(colorMap);

            Log.Trace("HasComponent - TransformComponent = {0}", HasComponent<TransformComponent>());
            Log.Trace("HasComponent - ScriptComponent = {0}", HasComponent<ScriptComponent>());
            Log.Trace("HasComponent - MeshComponent = {0}", HasComponent<MeshComponent>());

            MeshComponent meshComponent = GetComponent<MeshComponent>();
            if (meshComponent == null)
            {
                Log.Trace("MeshComponent is null!");
                meshComponent = CreateComponent<MeshComponent>();
            }
            meshComponent.Mesh = MeshFactory.CreatePlane(1.0f, 1.0f);

            Log.Trace("Mesh has {0} materials!", meshComponent.Mesh.GetMaterialCount());

            MaterialInstance material = meshComponent.Mesh.GetMaterial(1);
            material.SetKeyword("ALBEDO_MAP", true);
            material.Set("u_AlbedoTexture", texture);

            TransformComponent transformComponent = GetComponent<TransformComponent>();
            Vector3 position = Entity.GetTransform().Translation;
            transformComponent.Transform = Matrix4.Scale(new Vector3(10f, 1.0f, 10f)) * Matrix4.Translate(position);
        }

        private void OnCreate()
        {
            GenerateMap();
        }

        private void OnUpdate()
        {
            float ts = Time.DeltaTime;
            Matrix4 transform = Entity.GetTransform();
            Vector3 translation = transform.Translation;
            translation.Y += ts * Speed;
            if (Input.IsKeyPressed(KeyCode.Space))
            {
                translation.Y -= 10.0f;
            }
            transform.Translation = translation;
            Entity.SetTransform(transform);
        }
    }
}

File: PrismEditor/Assets/scripts/Python/MapGenerator.py
from Prism import (
    Behaviour, Log, Time, Input, KeyCodes,
    TransformComponent, MeshRendererComponent,
)
from Prism.Math import Vector2, Vector3, Vector4
from Prism.Math.Mathf import Mathf
from Prism.Renderer import Color, Texture2D, MeshFactory
import Noise as NoiseUtil


class MapGenerator(Behaviour):
    MapWidth: int = 128
    MapHeight: int = 128
    Octaves: int = 4
    Persistance: float = 0.74
    Seed: int = 21
    Lacunarity: float = 3.0
    Offset: Vector2 = Vector2(13.4, 6.26)
    NoiseScale: float = 0.5
    Speed: float = 0.0

    def GenerateMap(self):
        noiseMap = NoiseUtil.GenerateNoiseMap(
            self.MapWidth, self.MapHeight, self.Seed, self.NoiseScale,
            self.Octaves, self.Persistance, self.Lacunarity, self.Offset)

        width = len(noiseMap)
        height = len(noiseMap[0]) if width > 0 else 0

        texture = Texture2D(width, height)
        colorMap = []
        for y in range(height):
            for x in range(width):
                colorMap.append(Mathf.Lerp(Color.Black, Color.White, noiseMap[x][y]))

        texture.SetData(colorMap)

        Log.Trace("HasComponent - TransformComponent = {}", self.HasComponent(TransformComponent))
        Log.Trace("HasComponent - MeshRendererComponent = {}", self.HasComponent(MeshRendererComponent))

        meshComponent = self.GetComponent(MeshRendererComponent)
        if meshComponent is None:
            Log.Trace("MeshRendererComponent is null!")
            meshComponent = self.CreateComponent(MeshRendererComponent)

        meshComponent.Mesh = MeshFactory.CreatePlane(1.0, 1.0)

        Log.Trace("Mesh has {} materials!", meshComponent.GetMaterialCount())

        material = meshComponent.GetMaterial(1)
        material.SetKeyword("ALBEDO_MAP", True)
        material.SetTexture("u_AlbedoTexture", texture)

        transformComponent = self.GetComponent(TransformComponent)
        position = transformComponent.Position
        transformComponent.Scale = Vector3(10.0, 1.0, 10.0)
        transformComponent.Position = position

    def OnCreate(self):
        self.GenerateMap()

    def OnUpdate(self):
        ts = Time.DeltaTime
        pos = self.Entity.Transform.Position
        pos.y += ts * self.Speed
        if Input.IsKeyPressed(KeyCodes.Space):
            pos.y -= 10.0
        self.Entity.Transform.Position = pos
